Settings: XAML refactoring

Jay-o-Way opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

Overhaul of Settings & Welcome XAML files.

(invisible)
- Consolidated **namespaces** and (Visual Studio) removed unneeded namespaces
- Re-wrote bits of **x:bind** grammar: `{x:Bind Mode=m, Path=p}` → `{x:Bind p, Mode=m}`
- Fixed spelling: _EnableEtraBoxes_ → _EnableExtraBoxes_ (imagesize.cs)
- Re-wrote part on MouseUtils page where strings are concatenated
- A number of **ui:FontIcon** elements still had FontFamily and/or FontSize declared
- Trailing line endings

(visible)
- A few adjustments in **Margins** and **Spacings** - some visible, some not
- A few **Styles** (~incl. OobeSubtitleStyle → SubtitleTextBlockStyle~)
- OOBE: the **Hyperlink** gets a new line
- OOBE: module image is now centered
- removed incorrect background for OOBE Peek page


<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] Part of: #18866
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Localization:** All end user facing strings can be localized

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

[DataContext](https://github.com/microsoft/PowerToys/issues/18866#issuecomment-1804830273) -> new pr

![image](https://github.com/microsoft/PowerToys/assets/65828559/62ae3382-4227-44e2-9e72-575cf017f3f3)
↳ (0.75.1) This padding should not be here; missing space next to keys

![image](https://github.com/microsoft/PowerToys/assets/65828559/0d6b3a68-927e-47bd-a3b5-b3f56af7a3c8)
↳ (pr) ~SubtitleTextBlockStyle;~ Hyperlink on new line; padding fixed

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Build with Visual Studio.

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/AlphaColorPickerButton.xaml:
```diff
@@ -13,7 +12,7 @@
         <DropDownButton Padding="4,4,8,4" AutomationProperties.FullDescription="{x:Bind clr:ColorHelper.ToDisplayName(SelectedColor), Mode=OneWay }">
         -->
-        <DropDownButton Padding="4,4,8,4">
+        <DropDownButton Padding="4,4,12,4">
```
![image](https://github.com/microsoft/PowerToys/assets/65828559/b4ca34f5-75bb-4a7d-bd11-7598739e9d21)

htcfreek commented:
> OOBE : the Hyperlink gets a new line\n

What about switching order between hyperlink and buttons? 🤔 (First hyperlink, then buttins.)

Jay-o-Way commented:
Possible. But why? Links usually have a "less than important" feeling.

htcfreek commented:
> Possible. But why? Links usually have a "less than important" feeling.

I thought it might look better to have the "big" buttons after the "small" text and link, and not between.

htcfreek commented:
> Ready for review.
> I don't understand ☹️ How does `.\.pipelines\applyXamlStyling.ps1 -Main` work?

@jay-o-way
There is the VS plugin XamlStyler. You should install it and then use it to fix xaml code style.

Jay-o-Way commented:
Got it, thanks! Does it use default settings?

htcfreek commented:
> Got it, thanks! Assume it uses default settings?

Don't think so. There is a config in our repo: https://github.com/microsoft/PowerToys/blob/main/Settings.XamlStyler

jaimecbernardo commented:
Thanks for opening the PR!
Sorry for the delay Jay-o-Way, but thinking we want to get in all the WPFUI PR updates before the refactor goes in 😅

Jay-o-Way commented:
That's okay. I'm also looking forward to seeing those get in. I do have more ideas I want to work on, but to prevent issues I'll wait until all relevant pr's are handled...

jaimecbernardo reviewed (changes requested):
Hi, Thank you for opening the PR. I've given it a quick test and have got some crashes.
On PowerToys Run page, when I try to expand one of the plugins, it crashes the Settings app.
When I try to open the VCM Settings page, it crashes Settings.
Can you replicate these? Can you please look into them?

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/ColorFormatEditor.xaml:
```diff
@@ -36,11 +36,11 @@
         </DataTemplate>
 
-        <DataTemplate x:Key="ColorParameterTemplate" x:DataType="local:ColorFormatParameter">
+        <DataTemplate x:Key="ColorParameterTemplate" x:DataType="custom:ColorFormatParameter">
```
Maybe increase the spacing between the sections a bit.. either by increasing the spacing (red) or decrease (blue) the bottom part of each section
![image](https://github.com/microsoft/PowerToys/assets/9866362/95c63aee-c6d8-4ef3-beba-ce8ef767eb3a)

Jay-o-Way replied:
I knew you would mention this. 😄 My choice is based on...
![image](https://github.com/microsoft/PowerToys/assets/65828559/5fe00601-8a12-4fb8-8090-560b9ac80b57)

> Using Title, Subtitle and Body with 12epx spacing.

https://learn.microsoft.com/nl-nl/windows/apps/design/basics/content-basics#text--hierarchy

niels9001 replied:
I think the spacing is fine, but there is no hierarchy anymore. The title of a 'section' should provide more spacing so you know that the title/description belongs to the legend.

Jay-o-Way replied:
Maybe a Style would be better, then

niels9001 replied:
Just adding an additional top-margin to the titles/descriptions should work!

Jay-o-Way replied:
Sorry, your suggestion contradicts the guidance of Microsoft. It's confusing me and I'm not sure what to do, but my preference would be to follow the guidance.

Jay-o-Way replied:
> so you know that the title/description belongs to the legend.

I think the colon handles that.

niels9001 replied:
> Sorry, your suggestion contradicts the guidance of Microsoft. It's confusing me and I'm not sure what to do, but my preference would be to follow the guidance.

The information shown in this pop-up are sections, right? There needs to be more vertical spacing between sections vs. the title/label and content of a section. This helps to create a visual hierarchy. I don't believe the _guidelines_ you mentioned describe the spacing between sections?

Here's an example from the Fluent Windows Visual Library in Figma, which should be considered as our redlines for things. See how the spacing between the different sections is slightly higher (red) than the spacing between the header of the textbox (blue)?
<img width="604" alt="image" src="https://github.com/microsoft/PowerToys/assets/9866362/1aef6693-5e13-4c31-b5a2-ccbf2712e0a5">

niels9001 replied:
According to the design library, the spacing between sections (red) is 24px and spacing between header/child (blue) is 12px. Might be slightly different in XAML with the rendering of the TextBlock

Jay-o-Way replied:
Okay, let's take one step back: Do we consider the legend as a **control**, or as **text**? That's a big difference. [MS Learn](https://learn.microsoft.com/nl-nl/windows/apps/design/basics/content-basics#spacing-and-gutters) actually says to use **8px** between control and _its_ header, but looks clear on the 12px between text.
Also: I don't know Figma. If anything on MS Learn is outdated, that info should be updated. Otherwise I'm sticking to that.

niels9001 replied:
The docs are outdated here and there (as you can see your article is from 2021) - I think for this topic specifically, they are still correct.

In Figma it's 12px, but could very well be it's 8px in **e**px. It's a detail, the point is that the current spacing between the headers of the sections **does not provide the correct spacing**. 

For redlines and inspiration, the Windows Design team recommends to use the Fluent Windows Design Library - it's listed on the `Toolkits and libraries` page:
![image](https://github.com/microsoft/PowerToys/assets/9866362/8d031ff2-fa5b-43fd-bf0f-8596d4bc67c2)

niels9001 replied:
Looking at this, I wonder if we can make these titles less verbose?

"The following parameters:" => "**Parameters**"

"The red, green, blue and alpha values can be formatted to the following formats" => **"Formatting parameters"**

"Example: %ReX means red value in hex uppercase two digits format" => "Example: %ReX = ff"


Probably there are better ways to describe these :)

Jay-o-Way replied:
I've reset the visual to what it was.

Jay-o-Way replied:
I made it 12 and 8.

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/SettingsPageControl/SettingsPageControl.xaml:
```diff
@@ -19,23 +19,24 @@
             Text="{x:Bind ModuleTitle}" />
 
-        <ScrollViewer Grid.Row="1">
+        <ScrollViewer Grid.Row="1" Padding="36,0">
```
This might be due to these changes: on smaller sizes we want to decrease the spacing on the left / right to show more content:

Top is this PR, bottom is `main`
![image](https://github.com/microsoft/PowerToys/assets/9866362/7aae3e1e-912e-405b-ab67-ea0234b3114b)

Jay-o-Way replied:
> on smaller sizes we want to decrease the spacing on the left / right

Do you mean via a VisualState? I can look at that. Or simply decrease the "static" value? What are the best numbers then?

niels9001 replied:
I'm not sure if we want yet another visual state for this.. it was already working on `main` right? So maybe revert some of the padding changes you made - I think the values are the same but they now are e.g. margins instead of paddings

Jay-o-Way replied:
I reset the basic margin to 24. Check out the latest commit to see the VidualState. 🙂

niels9001 replied:
It works, nice 👍! 2 things:

Are you using 12 here? Could be that 16 is more appropriate. That would align the text with the navview button and would make the scrollbar not overlap the card itself?
![image](https://github.com/microsoft/PowerToys/assets/9866362/19c0f3c8-9584-4a85-acb9-0e92229f5cb4)

There seems to be a difference between `main` in this PR with the gap between the window border and the content?
![image](https://github.com/microsoft/PowerToys/assets/9866362/2c81dbdf-dce6-4e79-ba64-e2ba697b3599)

Jay-o-Way replied:
I'll look again

Jay-o-Way replied:
> That would align the text with the navview button and would make the scrollbar not overlap the card itself?

Yes, the scrollbar is a good argument. But Windows Settings does not align titlebar buttons with text/content below it. It uses 24px.

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/ShortcutControl/ShortcutDialogContentControl.xaml:
```diff
@@ -21,7 +24,7 @@
             Grid.Row="1"
             Height="56"
-            Margin="0,64,0,0"
+            Margin="0,40,0,0"
```
Something changed here, now when the `InfoBar` is shown the dialog "jumps" which isn't a great experience.

New vs. old
![Jumpingshortcutdialog](https://github.com/microsoft/PowerToys/assets/9866362/acb3c3d0-95d4-4440-b970-e4d0f4638911)

Jay-o-Way replied:
Ahh yes, I was wondering why the height was hard-coded... Maybe insert a grid _row_ for the InfoBar(s) with a MinHeight?

niels9001 replied:
Double check what there was originally and revert any changes there? (maybe it was the margin decrease?) - just to make sure we don't run into any other issues.

Jay-o-Way replied:
The row actually works well 🙂 check out the update?

niels9001 replied:
Works! But now the dialog height has increased? 

![image](https://github.com/microsoft/PowerToys/assets/9866362/2a9fba5f-7d6e-4a08-b6f2-d20d8e1b3c7d)

Jay-o-Way replied:
I'll tweak it

Jay-o-Way replied:
![image](https://github.com/microsoft/PowerToys/assets/65828559/db4a8bf3-ea90-4e37-9bdf-981a4417f1e6)
↳ Note: removing the **MinWidth** solves issue with text being cut off. It happens because the minimum with of the parent window and the ShortcutDialogContentControl were both **480**. Is it bad if the gutters are not exactly what they were?
The formatting of the tooltip text is off-topic, but good to know.

Jay-o-Way replied:
Off-topic: looks like the majority of MS app and Settings use a minwidth of 500.

niels9001 replied:
Oh yeah we can increase the MinWidth is needed.. it's up to the content of the dialog to see chose what value makes sense there. We just need to make sure that upon resizing things don't get clipped.

Dialog height increase is fine I guess.. the spacing between the infobar, the shortcut and text feels a bit off though?

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/OOBE/Views/OobeAlwaysOnTop.xaml:
```diff
@@ -4,31 +4,26 @@
         <custom:OOBEPageControl.PageContent>
-            <StackPanel Orientation="Vertical">
-                <TextBlock x:Uid="Oobe_HowToUse" Style="{ThemeResource OobeSubtitleStyle}" />
+            <StackPanel Orientation="Vertical" Spacing="12">
```
Top spacing is gone now? Also the subtitles are increased in size. 

New vs. old
![image](https://github.com/microsoft/PowerToys/assets/9866362/a04b8030-d2ee-42db-8f1e-0334a8516bd7)

niels9001 replied:
Spacing at the top needs to be reverted, subtitles as well.. this was done per design spec so we don't want to revert from that. There wasn't anything broken, right :)?

Jay-o-Way replied:
Is that design spec public? The current style looks the same as headers for settings (groups) but this isn't settings.

niels9001 replied:
I don't think it is, we were following an internal draft of an OOBE page. But since we didn't get any feedback on this specifically (from users or internal teams) maybe we should leave this as is? We are thinking of a bigger overhaul of the entire OOBE experience, will post an update in the issue once I have that

Jay-o-Way replied:
Are you okay with using this style though? Since I can't know the design spec you are referring to, and it does add a bit of hierarchy.

niels9001 replied:
The spacing was already providing the hierarchy - and these titles are not the primary to convery, especially since they are the same on every page. Therefore they wouldn't need additional emphasis.

The additional spacing between the buttons at the bottom and the text has changed now as well? Might want to increase that, see earlier comment.

Jay-o-Way replied:
I've reset the visual to what it was.

niels9001 reviewed (changes requested):
See comments

niels9001 commented:
> Hi, Thank you for opening the PR. I've given it a quick test and have got some crashes. On PowerToys Run page, when I try to expand one of the plugins, it crashes the Settings app. When I try to open the VCM Settings page, it crashes Settings. Can you replicate these? Can you please look into them?

@jaimecbernardo All plugins? I have tried a few and I'm not seeing any crashes 😲. This was running `PowerToys.Settings` in debug mode from VS

Jay-o-Way commented:
I have some build errors now, that were not present a while ago. They boil down to **GPOWrapper.cs** and **WinRT.cs**

htcfreek commented:
> @jaimecbernardo I cannot reproduce the issues you mentioned.
> I _do_ have some (re)build errors now, that were not present a while ago. They boil down to **GPOWrapper.cs** and **WinRT.cs**

Aaarg. That's a build thing that the settings page not automatically rebuilds all dependencies after tha changed. Please do a full "Build" instead of a "Rebuild" of the Settings project. And don't simply per the "Debug" button.

Jay-o-Way commented:
@htcfreek I did. Twice already.

htcfreek commented:
@Jay-o-Way 
Did you delete a comment/answer?

Jay-o-Way commented:
Yeah, errors are gone :)

Jay-o-Way commented:
PTAL

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/SettingsPageControl/SettingsPageControl.xaml:
```diff
@@ -27,16 +27,17 @@
             x:Name="Header"
             MaxWidth="{StaticResource PageMaxWidth}"
-            Padding="36,0"
+            Padding="24,0"
```
@jay-o-way What is the advantage of all of these changes here? Seems the XAML is becoming more complex but not visual changes are intended?

Jay-o-Way replied:
Not sure if this answers your question, but I guess my brain has a different approach for implementing these things?

niels9001 replied:
I just want to make sure we are not breaking things that doesn't need fixing :)! If you can verify that there are no regressions (on different window sizes) this is fine!

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/SettingsPageControl/SettingsPageControl.xaml:
```diff
@@ -46,7 +47,10 @@
                 <!--  Top panel  -->
-                <Grid MaxWidth="{StaticResource PageMaxWidth}" ColumnSpacing="16">
+                <Grid
+                    x:Name="TopGrid"
```
I think we want to avoid naming things that we don't reference in code-behind. Next time someone wants to edit this code, it feels like I shouldn't touch it because there might be C# referencing this?

Jay-o-Way replied:
This was to use with the VisualState. Should it be `Name` instead?

jaimecbernardo commented:
I usually run this incantation when switching branches:
`git clean -ffxd`

So, Settings doesn't crash for me if I run it directly from Visual Studio (PowerToys Run plugins, at least, since VCM page is disabled when running like that).

It still crashes if I run it through "Start without Debugging"(Ctrl+F5) or if I start the runner and then start Settings from the tray icon. (This is what I usually do when testing since it's the normal way to start the Settings apps)
>On PowerToys Run page, when I try to expand one of the plugins, it crashes the Settings app.
When I try to open the VCM Settings page, it crashes Settings.

Does it crash for you as well?

jaimecbernardo commented:
I'm able to get the following stack traces by analyzing the stowed exception in full crash dumps, in case it helps:

For VCM, the message error is:
```
The parameter is incorrect.

The value cannot be converted to type ImageSource
```
Stack trace:
```
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|39_0(Int32 hr)
   at ABI.Microsoft.UI.Xaml.Markup.IXamlBindingHelperStaticsMethods.ConvertValue(IObjectReference _obj, Type type, Object value) in D:\a\_work\1\s\BuildOutput\obj\ARM64fre\src\projection\generated\Microsoft.UI.Xaml.Markup.cs:line 1723
   at Microsoft.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(Type type, Object value) in D:\a\_work\1\s\BuildOutput\obj\ARM64fre\src\projection\generated\Microsoft.UI.Xaml.Markup.cs:line 624
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Update_ViewModel_CameraImageOverlayPath(String obj, Int32 phase) in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 676
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Update_ViewModel(VideoConferenceViewModel obj, Int32 phase) in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 506
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Update_(VideoConferencePage obj, Int32 phase) in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 480
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Update() in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 421
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Initialize() in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 415
   at Microsoft.PowerToys.Settings.UI.Views.VideoConferencePage.VideoConferencePage_obj1_Bindings.Loading(FrameworkElement src, Object data) in C:\prog\janeasystems\PowerToys\src\settings-ui\Settings.UI\obj\x64\Debug\SettingsXAML\Views\VideoConference.g.cs:line 454
   at WinRT._EventSource_global__Windows_Foundation_TypedEventHandler_global__Microsoft_UI_Xaml_FrameworkElement__object_.EventState.<GetEventInvoke>b__1_0(FrameworkElement sender, Object args) in D:\a\_work\1\s\BuildOutput\obj\ARM64fre\src\projection\generated\WinRTEventHelpers.cs:line 2523
   at ABI.Windows.Foundation.TypedEventHandler`2.Do_Abi_Invoke[TSenderAbi,TResultAbi](Void* thisPtr, TSenderAbi sender, TResultAbi args)
```

For PowerToys Run plugins, the error message is :
```
No installed components were detected.

Cannot apply a Style with TargetType 'CommunityToolkit.WinUI.Controls.SettingsCard' to an object of type 'Microsoft.UI.Xaml.Controls.InfoBar'.
```
Stack trace:
```
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|39_0(Int32 hr)
   at ABI.Microsoft.UI.Xaml.IFrameworkElementMethods.set_Style(IObjectReference _obj, Style value) in D:\a\_work\1\s\BuildOutput\obj\ARM64fre\src\projection\generated\Microsoft.UI.Xaml.cs:line 17520
   at CommunityToolkit.WinUI.Controls.SettingsExpander.ItemsRepeater_ElementPrepared(ItemsRepeater sender, ItemsRepeaterElementPreparedEventArgs args)
   at WinRT._EventSource_global__Windows_Foundation_TypedEventHandler_global__Microsoft_UI_Xaml_Controls_ItemsRepeater__global__Microsoft_UI_Xaml_Controls_ItemsRepeaterElementPreparedEventArgs_.EventState.<GetEventInvoke>b__1_0(ItemsRepeater sender, ItemsRepeaterElementPreparedEventArgs args) in D:\a\_work\1\s\BuildOutput\obj\ARM64fre\src\projection\generated\WinRTEventHelpers.cs:line 1931
   at ABI.Windows.Foundation.TypedEventHandler`2.Do_Abi_Invoke[TSenderAbi,TResultAbi](Void* thisPtr, TSenderAbi sender, TResultAbi args)
```

Really love how both stack trace seem to refer to ARM64 stuff? That's pretty weird.

niels9001 commented:
As I mentioned and highlighted in my earlier comments, this PR contains a few visual changes decreasing the whitespace between textblocks and overall UI elements. This decreases the readability, accesibility and therefore degrading the user experience.
Properly spaced elements and textblocks are important to let the reader follow and understand what belongs to what. It creates structural clarity and provides emphasis and focus on what really matters. With incorrect whitespacing, your eye needs to scan the page to understand what belongs to what which could potentially result in eye strain or fatigue. The visual hierarchy (with the help of whitespacing) helps with this. If you want to learn more about this fundemental design principle, check out: https://www.angelovanikoletadesign.com/post/how-to-effectively-use-whitespace-in-design-for-better-readability and https://verpex.com/blog/website-tips/what-is-whitespace-and-why-is-it-important

Updating values based on docs is fine, but please treat these as guidelines and not UI redlines - these are open for interpretation and do not cover all scenarios. Or try to bring these things up in (seperate!) issues so we can discuss these before doing the work and we don't need to spend this much time reviewing.

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/ColorFormatEditor.xaml:
```diff
@@ -101,9 +96,8 @@
 
             <TextBlock
```
Looking at this, I wonder if we can make these titles less verbose?

"The following parameters:" => "Parameters"

"The red, green, blue and alpha values can be formatted to the following formats" => "Formatting parameters"

"Example: %ReX means red value in hex uppercase two digits format" => "Example: %ReX = ff"

Probably there are better ways to describe these :)

Jay-o-Way replied:
Noted

jaimecbernardo reviewed (changes requested):
> I usually run this incantation when switching branches: `git clean -ffxd`
> 
> So, Settings doesn't crash for me if I run it directly from Visual Studio (PowerToys Run plugins, at least, since VCM page is disabled when running like that).
> 
> It still crashes if I run it through "Start without Debugging"(Ctrl+F5) or if I start the runner and then start Settings from the tray icon. (This is what I usually do when testing since it's the normal way to start the Settings apps)
> 
> > On PowerToys Run page, when I try to expand one of the plugins, it crashes the Settings app.
> > When I try to open the VCM Settings page, it crashes Settings.
> 
> Does it crash for you as well?

I gave the PR a try again and it still crashes for me on these two places.

Jay-o-Way commented:
> I gave the PR a try again and it still crashes for me on these two places.

The only issue on the page, that I can find now, is 
![image](https://github.com/microsoft/PowerToys/assets/65828559/8cd203b5-1508-484b-a011-3b190df743e7)
I'll investigate... 🕵🏻

htcfreek commented:
@Jay-o-Way , @jaimecbernardo 
Did you both try with a clean build or did one of you only do rebuilds?

Jay-o-Way commented:
The error is independent of a branch - even `main`. Therefore, building the project is a problem in any case.

Jay-o-Way commented:
@jaimecbernardo @niels9001 GREAT NEWS!
I have found and fixed the cause of the error. I had changed a `SettingsCard + InfoBar` to `InfoBar`, but that wasn't allowed.

Jay-o-Way commented:
Hey @stefansjfw, I see you're active. Would you like to give an extra look (aka. review) at this, or would you rather leave it up to Jaime and Niels? In any case, happy holidays 🎄🎆

Jay-o-Way commented:
Guys, have a moment? 😇

niels9001 commented:
> Guys, have a moment? 😇

Most of us are out of office for the holidays. I'll make sure to get back to this once I'm back!

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Views/EnvironmentVariablesPage.xaml:
```diff
@@ -2,51 +2,47 @@
     xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
-    xmlns:controls="using:Microsoft.PowerToys.Settings.UI.Controls"
-    xmlns:controls1="using:CommunityToolkit.WinUI.Controls"
+    xmlns:custom="using:Microsoft.PowerToys.Settings.UI.Controls"
```
@Jay-o-Way Why is this namespace `custom` now? That's rather confusing actually, since they are using the `.Controls` namespace? That's why we use `converters` namespace for Converters

Can't we use `controls` for PowerToys controls, and `toolkit` or `toolkitcontrols` for all controls part of the Toolkit?

niels9001 replied:
Actually, looking at the updated XAML.. `toolkitcontrols` is pretty verbose and makes the XAML harder to read :(.

Could we use `tk:SettingsCard` instead maybe? Or we do use `custom` for custom controls (it's used only in a handful of files) and use `controls` for Toolkit controls?

Jay-o-Way replied:
> Why is this namespace `custom` now?

Because... In the `main` branch, `custom` is already (consistently) used through-out the OOBE xaml pages. Under the rest of _SettingsXAML_ area/folder, `controls`, `local`, and `custom` are all used. Just let me know which one is the best and I will apply to all of them. *
![image](https://github.com/microsoft/PowerToys/assets/65828559/0b363366-a0e8-4417-b1d5-14095851c03f)

https://github.com/microsoft/PowerToys/issues/18866#issuecomment-1806326170

> looking at the updated XAML.. `toolkitcontrols` is pretty verbose and makes the XAML harder to read :(.
> 
> Could we use `tk:SettingsCard` instead maybe?

Sure, that's possible. I would suggest `tkc` though, to include the "C" for controls. \* In that light: `ptc` for PowerToys Controls.

Jay-o-Way replied:
> Why is this namespace `custom` now?

and

> looking at the updated XAML.. `toolkitcontrols` is pretty verbose and makes the XAML harder to read :(.
> 
> Could we use `tk:SettingsCard` instead maybe?In the `main` branch, `custom` is already (consistently) used through-out the OOBE xaml pages. Under the rest of _SettingsXAML_ area/folder, `controls`, `local`, and `custom` are all used.

@niels9001 Both **PowerToys** and **ToolKit** can have namespaces for **Controls** and **Converters** declared. That's four combinations. Simply using "controls" or "converters" is ambiguous. We can shorten "toolkit" to "tk" and "PowerToys" to "pt", but when both versions of **Controls** or **Converters** are used in the same file, we can't shorten it any more than that. (Aside from skipping vowels...)

Therefore I suggest

- xmlns:ptcontrols="using:Microsoft.PowerToys.Settings.UI.Controls"
- xmlns:ptconverters="using:Microsoft.PowerToys.Settings.UI.Converters"
- xmlns:tkcontrols="using:CommunityToolkit.WinUI.UI.Controls"
- xmlns:tkconverters="using:CommunityToolkit.WinUI.Converters"

https://github.com/microsoft/PowerToys/issues/18866#issuecomment-1806326170

Images:

![image](https://github.com/microsoft/PowerToys/assets/65828559/0b363366-a0e8-4417-b1d5-14095851c03f)
![image](https://github.com/microsoft/PowerToys/assets/65828559/96f3e861-e177-4dd0-b583-3cf3428a378f)
![image](https://github.com/microsoft/PowerToys/assets/65828559/8f8243ab-4842-4331-8e9e-8aaef28ba87a)

Jay-o-Way commented:
@niels9001 you're right. The longer it takes, the bigger it gets. Maybe I should re-do this into a few prs.

- Namespaces (once we settle on a solution, see [latest review comments](https://github.com/microsoft/PowerToys/pull/29785#discussion_r1438328130))
- Bindings
- Visual improvements
- etc.
